smolyay/test_function_class.py

```diff
@@ -0,0 +1,6478 @@
+import abc
+
+import numpy
+
+class test_fun(abc.ABC):
+    """Abstract test function class
+
+    Stores data about each test function including call, name, the number of
+    dimensions, the domain, and how to normalize that domain to the interval
+    [-1,1]
+
+    Parameters
+    ----------
+    name : str
+        the name of the class
+
+    dim : int
+        the number of variables used as input
+
+    lower_bounds : list of floats of size 1xdim
+        the lower bounds of the domain of each variable
+
+    upper_bounds : list of floats of size 1xdim
+        the upper bounds of the domain of each variable
+
+    domain : list of lists
+        the domain of the inputs for each dimension
+    """
+    def __call__(self,x):
+        """Evaluate the test function
+
+        Parameters
+        ----------
+        x : list of floats
+            the input to the call method
+
+        Raises
+        ------
+        ValueError
+            if the input is outside is outside the domain
+        """
+        if self.dim > 1:
+            oob = any(
+                xi < bound[0]-1e-10 or xi > bound[1]+1e-10 for xi,bound in
```

Comment:
The domain is closed, so you can compare using < and >. If you are getting an error elsewhere, that is an issue with how that value of x is being generated.
```suggestion
                xi < bound[0] or xi > bound[1] for xi,bound in
```

Reply:
The original reason that I put that there was because there were rare occasions in which while trying to get the samples for training the surrogate model, the function would give the out of bounds error. The issue appears to be rounding from floating point arithmetic in the _mapdomain function in the surrogate class.

For example, trying to make a surrogate function for the BenchmarkFunction ex8_1_3, the surrogate wants to sample from the point (8.460000000000003, -0.8800000000000008) but the domain of the first variable is [-10.6, 8.46].

Reply:
I see. We should then take a look at that method and see if we can improve it in some way. One option would be to clamp the values to lie within the domain.

Reply:
I tried to change the _mapdomain function to see if I could force the bounds using numpy.interp. However, the solution I came up with that passes all the pytests is very messy.

It replaces the original return statement with this:
`x = numpy.array(x, copy=False, ndmin=1)`
`new_x = numpy.zeros(x.shape)`
`if not 1 in old.shape:`
`    # we have more than one domain`
`    if not len(x.shape) == 1:`
`        # many values in x share a domain`
`        for i in range(x.shape[-1]):`
`            new_x[:,i] = numpy.interp(x[:,i],old[i,:],new[i,:])`
`    else:`
`        # each value in x use different domains`
`        for i in range(x.shape[-1]):`
`            new_x[i] = numpy.interp(x[i],old[i,:],new[i,:])`
`else:`
`    # all values in x use the same domain`
`    new_x = numpy.interp(x,old[0],new[0])`
`return new_x`

numpy.interp keeps everything in the bounds just fine, but it's very picky about the shape of old and new, and it doesn't appear to want to do multiple domains at once.

Reply:
How about something simpler like this?
```
# clamp lower bound
flags = new_x < new[:, 0]
new_x[flags] = new[flags, 0]

# clamp upper bound
flags = new_x > new[:, 1]
new_x[flags] = new[flags, 1]
```

Reply:
Yeah that's much simpler.